File: GymPlatform/e2e/src/test/java/com/gymplatform/e2e/pages/PosPage.java
package com.gymplatform.e2e.pages;

import com.gymplatform.e2e.support.SeleniumHelpers;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class PosPage {

    private final WebDriver driver;
    private final WebDriverWait wait;

    public PosPage(WebDriver driver, WebDriverWait wait) {
        this.driver = driver;
        this.wait = wait;
    }

    public void open(String webBase) {
        driver.get(webBase + "/ventas/punto-de-venta");
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".pos-catalog, .pos-toolbar")));
    }

    public boolean isCashOpen() {
        return driver.findElement(By.cssSelector(".pos-cash-pill")).getText().contains("Caja abierta");
    }

    public void dismissBlockingOverlays() {
        var summaryClose = driver.findElements(
                By.cssSelector(".pos-sale-summary button[aria-label='Cerrar resumen']"));
        if (!summaryClose.isEmpty() && summaryClose.get(0).isDisplayed()) {
            SeleniumHelpers.jsClick(driver, summaryClose.get(0));
        }
        if (!driver.findElements(By.cssSelector(".confirm-dialog")).isEmpty()) {
            new ConfirmDialogPage(driver, wait).cancel();
        }
    }

    public void clickCloseCashRegister() {
        dismissBlockingOverlays();
        WebElement btn = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//button[normalize-space()='Cerrar caja' and not(@disabled)]")));
        SeleniumHelpers.jsClick(driver, btn);
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".cash-session-modal")));
    }

    public void clickOpenCashRegister() {
        WebElement btn = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//button[normalize-space()='Abrir caja' and not(@disabled)]")));
        SeleniumHelpers.jsClick(driver, btn);
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".cash-session-modal")));
    }

    public void ensureCashOpen(CashSessionPage cash) {
        if (!isCashOpen()) {
            clickOpenCashRegister();
            cash.openWithDefaultFloat();
        }
        driver.navigate().refresh();
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".pos-catalog, .pos-toolbar")));
        if (!isCashOpen()) {
            clickOpenCashRegister();
            cash.openWithDefaultFloat();
        }
        wait.until(ExpectedConditions.textToBePresentInElementLocated(
                By.cssSelector(".pos-cash-pill"), "Caja abierta"));
        wait.until(driver -> !driver.findElements(By.cssSelector("button.pos-product-main:not([disabled])")).isEmpty());
    }

    public void ensureProductsCatalogTab() {
        WebElement tab = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//div[contains(@class,'pos-catalog-head')]//button[contains(@class,'product-manage-tab') and contains(.,'Productos')]")));
        if (!tab.getAttribute("class").contains("active")) {
            SeleniumHelpers.jsClick(driver, tab);
        }
    }

    public void addFirstProductToCart() {
        ensureProductsCatalogTab();
        for (WebElement card : driver.findElements(By.cssSelector(".pos-product-card:not(.is-sold-out)"))) {
            var buttons = card.findElements(By.cssSelector("button.pos-product-main:not([disabled])"));
            if (buttons.isEmpty()) {
                continue;
            }
            SeleniumHelpers.jsClick(driver, buttons.get(0));
            if (!driver.findElements(By.cssSelector(".confirm-dialog")).isEmpty()) {
                new ConfirmDialogPage(driver, wait).confirm("Sí, vender contenedor");
            }
            wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".pos-cart-line")));
            return;
        }
        throw new IllegalStateException("No hay productos vendibles con caja abierta");
    }

    public void checkoutAllCash() {
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".pos-cart-line")));
        WebElement cobrar = wait.until(ExpectedConditions.elementToBeClickable(
                By.cssSelector("aside.pos-cart button.pos-checkout-btn:not([disabled])")));
        SeleniumHelpers.jsClick(driver, cobrar);
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".pos-payment-modal")));
        WebElement allCash = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//div[contains(@class,'pos-payment-modal')]//button[normalize-space()='Todo efectivo']")));
        SeleniumHelpers.jsClick(driver, allCash);
        WebElement confirm = wait.until(ExpectedConditions.elementToBeClickable(
                By.xpath("//div[contains(@class,'pos-payment-modal')]//button[normalize-space()='Confirmar cobro' and not(@disabled)]")));
        SeleniumHelpers.jsClick(driver, confirm);
        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector(".pos-payment-modal")));
    }
}

File: GymPlatform/web/cypress/support/pages/pos.page.ts
import * as cash from './cash-session.page'

export function visitPos() {
  cy.visit('/ventas/punto-de-venta')
  cy.get('.pos-catalog, .pos-toolbar').should('be.visible')
}

export function isCashOpen() {
  cy.get('.pos-cash-pill.is-open').should('exist')
}

export function dismissBlockingOverlays() {
  cy.get('body').then(($body) => {
    const summaryClose = $body.find(".pos-sale-summary button[aria-label='Cerrar resumen']")
    if (summaryClose.length && summaryClose.is(':visible')) {
      cy.wrap(summaryClose.first()).click({ force: true })
    }
  })
  cy.get('body').then(($body) => {
    if ($body.find('.confirm-dialog').length) {
      cy.contains('.confirm-dialog button', 'Cancelar').click()
    }
  })
}

export function clickOpenCashRegister() {
  cy.contains('button', 'Abrir caja').should('not.be.disabled').click()
  cash.waitForCashModal('Abrir caja')
}

export function clickCloseCashRegister() {
  dismissBlockingOverlays()
  cy.contains('button', 'Cerrar caja').should('not.be.disabled').click()
  cash.waitForCashModal('Cerrar caja')
}

export function ensureCashOpen() {
  cy.get('.pos-cash-pill').then(($pill) => {
    if (!$pill.hasClass('is-open')) {
      clickOpenCashRegister()
      cash.openWithDefaultFloat()
    }
  })
  cy.reload()
  cy.get('.pos-catalog, .pos-toolbar').should('be.visible')
  cy.get('.pos-cash-pill.is-open').should('exist')
  cy.get('button.pos-product-main:not([disabled])').should('have.length.at.least', 1)
}

export function ensureProductsTab() {
  cy.get('.pos-catalog-head').contains('button.product-manage-tab', 'Productos').then(($tab) => {
    if (!$tab.hasClass('active')) {
      cy.wrap($tab).click()
    }
  })
}

export function addProductToCart(productName: string) {
  ensureProductsTab()
  cy.contains('.pos-product-card', productName).within(() => {
    cy.get('button.pos-product-main:not([disabled])').click()
  })
  cy.get('body').then(($body) => {
    if ($body.find('.confirm-dialog').length) {
      cy.confirmDialog('Sí, vender contenedor')
    }
  })
  cy.get('.pos-cart-line').should('be.visible')
}

export function assertProductSoldOutInPos(productName: string) {
  cy.contains('.pos-product-card', productName).should('have.class', 'is-sold-out')
  cy.contains('.pos-product-card', productName).find('button.pos-product-main').should('be.disabled')
}

export function openCheckout() {
  cy.get('aside.pos-cart button.pos-checkout-btn:not([disabled])').click()
  cy.get('.pos-payment-modal').should('be.visible')
}

export function checkoutAllCash() {
  openCheckout()
  cy.contains('.pos-payment-modal button', 'Todo efectivo').click()
  cy.contains('.pos-payment-modal button', 'Confirmar cobro').should('not.be.disabled').click()
  cy.get('.pos-payment-modal').should('not.exist')
}

export function checkoutAllSinpe() {
  openCheckout()
  cy.contains('.pos-payment-modal button', 'Todo SINPE').click()
  cy.contains('.pos-payment-modal button', 'Confirmar cobro').should('not.be.disabled').click()
  cy.get('.pos-payment-modal').should('not.exist')
}

export function checkoutMixedCashSinpe(cashAmount: number, sinpeAmount: number) {
  openCheckout()
  cy.contains('.pos-payment-modal label', 'Efectivo (₡)').parent().find('input').clear().type(String(cashAmount))
  cy.contains('.pos-payment-modal label', 'SINPE (₡)').parent().find('input').clear().type(String(sinpeAmount))
  cy.contains('.pos-payment-modal button', 'Confirmar cobro').should('not.be.disabled').click()
  cy.get('.pos-payment-modal').should('not.exist')
}

export function closeCashMatchingExpected() {
  clickCloseCashRegister()
  cash.closeMatchingExpected()
  cy.get('.pos-cash-pill:not(.is-open)').should('exist')
}
